keys/signer_secp256r1.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 Coinbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package keys
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+
+	"github.com/coinbase/rosetta-sdk-go/asserter"
+	"github.com/coinbase/rosetta-sdk-go/types"
+)
+
+// SignerSecp256r1 is initialized from a keypair
+type SignerSecp256r1 struct {
+	KeyPair *KeyPair
+}
+
+// The Ecdsa signature is the couple (r, s), both r and s are 32 bytes
+const (
+	EcdsaRLen = 32
+	EcdsaSLen = 32
+)
+
+var _ Signer = (*SignerSecp256r1)(nil)
+
+// PublicKey returns the PublicKey of the signer
+func (s *SignerSecp256r1) PublicKey() *types.PublicKey {
+	return s.KeyPair.PublicKey
+}
+
+// Sign arbitrary payloads using a KeyPair
+func (s *SignerSecp256r1) Sign(
+	payload *types.SigningPayload,
+	sigType types.SignatureType,
+) (*types.Signature, error) {
+	if err := s.KeyPair.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if !(payload.SignatureType == sigType || payload.SignatureType == "") {
+		return nil, fmt.Errorf(
+			"%w: %v",
+			ErrSignUnsupportedPayloadSignatureType,
+			payload.SignatureType,
+		)
+	}
+
+	if sigType != types.Ecdsa {
+		return nil, fmt.Errorf(
+			"%w: expected %v but got %v",
+			ErrSignUnsupportedSignatureType,
+			types.Ecdsa,
+			sigType,
+		)
+	}
+
+	crv := elliptic.P256()
+	x, y := crv.ScalarBaseMult(s.KeyPair.PrivateKey)
+	pubKey := ecdsa.PublicKey{X: x, Y: y, Curve: crv}
+	privKey := ecdsa.PrivateKey{
+		PublicKey: pubKey,
+		D:         new(big.Int).SetBytes(s.KeyPair.PrivateKey),
+	}
+
+	sigR, sigS, err := ecdsa.Sign(rand.Reader, &privKey, payload.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrSignFailed, err.Error())
+	}
+	sig := sigR.Bytes()
```

Comment:
check signature is not identity otherwise it will always be valid.

Reply:
This is handled by `ecdsa.Sign` method, please check: https://github.com/golang/go/blob/5a25a3fd1d8248c039e34f5ae02cc9c8198fc6d6/src/crypto/ecdsa/ecdsa.go#L262

Reply:
No that's not it. The signature is composed of two points, R and S. You should check that neither of these is equal to zero aka the identity element on the curve. If the signature is the identity element, it will always be valid.

Reply:
The check for S is here: https://github.com/golang/go/blob/5a25a3fd1d8248c039e34f5ae02cc9c8198fc6d6/src/crypto/ecdsa/ecdsa.go#L272

Reply:
Yeah I think you're right, it seems to retry if R or S is 0, but will wait for Mike to ack here.

Reply:
@mikelodder7 can you re-review?